### DISCUSSION TITLE
Computing SHA for use in CreateFile endpoint

### DISCUSSION THREAD

**Discussion by kevinburke** [community/community#24445] [Apps, API and Webhooks]
<p>I have a file whose contents are “from test” - 9 bytes. The documentation says that the SHA of created files is computed using SHA1:</p>
<p>&gt; The file’s SHA-1 hash is computed and stored in the blob object.</p>
<p>(from <a href="https://developer.github.com/v3/git/blobs/" rel="nofollow noopener">https://developer.github.com/v3/git/blobs/</a>)</p>
<p>However, when I compute the hex-encoded SHA1 output of “from test”, I get 5669556d9a5c27fdd649dcaaa0873757c2aa402f.</p>
<p>The Github API says that the sha is 62b551731eada762035d4665978027cd44291290 - this is the ETag returned and the value for “sha” in the API response for retrieving a file. In addition, when I call the CreateFile endpoint with “from test” as the value and 566955… as the sha, I’m told that the SHA is incorrect.</p>
<p>I’ve also tried appending newlines, computing the SHA of the base64 encoded value of the content, computing the SHA of base64+ a trailing newline, and none of them give me 62b551731eada762035d4665978027cd44291290. How is Github computing that value?</p>
<p>Thanks,</p>
<p>Kevin</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>GitHub uses git to calculate the SHA of the object. You can find <a href="https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#_object_storage" rel="nofollow noopener">the documentation of how this works</a> on the <a href="http://git-scm.com" rel="nofollow noopener">git-scm.com</a> website. Or you can use the <code>git hash-object</code> command:</p>
<pre><code>$ echo -n "from test" | git hash-object --stdin
62b551731eada762035d4665978027cd44291290
</code></pre>
<p>I hope that helps!</p>
